File: cs12m/lab7/Dictionary.java
public class Dictionary implements DictionaryInterface{

// NodeObj
private class Node{
   String key;
   String value;
   Node left;
   Node right;

// newNode()
// constructor for private Node type
Node (String k, String v) {
   key = k;
   value = v;
   left = null;
   right = null;
   }
}

private Node root;
private int numPairs;

// DictionaryObj
Dictionary(){
   Node root = null;
   int numPairs = 0;
} 

// findKey()
// returns the Node containing key k in the subtree rooted at R, or returns
// NULL if no such Node exists
private Node findKey(Node R, String k){
   if(R== null || k == R.key) 
      return R;
   if(k.compareTo(R.key) < 0 ){ 
      return findKey(R.left, k);
   }else{
      return findKey(R.right, k);
   } 
}

// findParent()
// returns the parent of N in the subtree rooted at R, or returns NULL 
// if N is equal to R. (pre: R != NULL)
private Node findParent(Node N, Node R){
   Node P=null;
   if( N!=R ){
      P = R;
      while( P.left!=N && P.right!=N ){
         if(N.key.compareTo(P.key) <0)
            P = P.left;
         else
            P = P.right;
      }
   }
   return P;
}

// findLeftmost()
// returns the leftmost Node in the subtree rooted at R, or NULL if R is NULL.
private Node findLeftmost(Node R){
   Node L = R;
   if( L!=null ) for( ; L.left!=null; L=L.left) ;
   return L;
}

// printInOrder()
// prints the (key, value) pairs belonging to the subtree rooted at R in order
// of increasing keys to file pointed to by out.
private String printInOrder(Node R){
   String s = "";
   if( R!=null ){
      s = printInOrder(R.left)+R.key+" "+ R.value+"\n"+printInOrder(R.right);
   }
   return s; 
}

// deleteAll()
// deletes all Nodes in the subtree rooted at N.
private void deleteAll(Node N){
   if( N!=null ){
      deleteAll(N.left);
      deleteAll(N.right);
   }
}

// public functions -----------------------------------------------------------

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
public boolean isEmpty(){
   return(numPairs==0);
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
public int size(){
   return numPairs;
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none
public String lookup(String key){
   Node N=findKey(root,key);
   if( N != null ){
     return N.value;
   }else{
     return null;
  } 
}
// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
public void insert(String k, String v) throws KeyCollisionException{
   Node N, A, B;
   if( findKey(root, k)!=null){
      throw new KeyCollisionException("Dictionary error: Cannot insert duplcate keys.");  
   }
   N = new Node(k, v);
   B = null;
   A = root;
   while( A!=null ){
      B = A;
      if(k.compareTo(A.key) <0) A = A.left;
      else A = A.right;
   }
   if( B==null ) root = N;
   else if(k.compareTo(B.key) <0) B.left = N;
   else B.right = N;
   numPairs++;
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
public void delete(String k) throws KeyNotFoundException{
   Node N, P, S;
   N = findKey(root, k);
   if( N== null ){
         throw new KeyCollisionException("Dictionary Error: cannot delete() non-existent key: ");
   }
   if( N.left==null && N.right==null ){
      if( N==root ){
         root = null;
      }else{
         P = findParent(N, root);
         if( P.right==N ) P.right = null;
         else P.left = null;
      }
   }else if( N.right== null){
      if( N==root ){
         root = N.left;
      }else{
         P = findParent(N, root);
         if( P.right==N ) P.right = N.left;
         else P.left = N.left;
      }
   }else if( N.left==null ){
      if( N==root ){
         root = N.right;
      }else{
         P = findParent(N,root);
         if( P.right==N ) P.right = N.right;
         else P.left = N.right;
      }
   }else{  // N->left!=NULL && N->right!=NULL
      S = findLeftmost(N.right);
      N.key = S.key;
      N.value = S.value;
      P = findParent(S, N);
      if( P.right==S ) P.right = S.right;
      else P.left = S.right;
   }
   numPairs--;
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
public void makeEmpty(){
   root = null;
   numPairs = 0;
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
public String toString(){
   String s = printInOrder(root);
   return s;
   }
}

File: cs12m/lab7/Dictionary.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"


// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* x, char* y){
	Node N = malloc(sizeof(NodeObj));
	assert(N!=NULL);
	N->key = x;
	N->value = y;
	//N->key = calloc(strlen(x),sizeof(char));
	//sprintf(N->key, "%s\0", x);
	//N->value = calloc(strlen(y),sizeof(char)); 
	//sprintf(N->value, "%s\0", y);
	N->next = NULL;
	return(N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN){
	if( pN!=NULL && *pN!=NULL ){
		Node N = *pN;
		//free(N->key);
		//free(N->value);
		free(N);
		*pN = NULL;
	}
}

// DicitonaryObj
typedef struct DictionaryObj{
	Node head;
	int numItems;
} DictionaryObj;

// public functions -----------------------------------------------------------

Node findKey(Dictionary D, char* x){
	Node N = D->head;
	while(N != NULL){
		if(strcmp(N->key, x) == 0){
			return N;
		}else{
			N = N->next;
		}
	}
	return NULL;    
}


// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){
	Dictionary D = malloc(sizeof(DictionaryObj));
	assert(D!=NULL);
	D->head = NULL;
	D->numItems = 0;
	return D;
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
	if( D==NULL ){
		fprintf(stderr, 
				"Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	return(D->numItems==0);

}

// lookup()
// returns (key value) specified from D
// pre: !isEmpty(S)
char* lookup(Dictionary D, char* x){
	if(D == NULL){
		fprintf(stderr, "Dictionary Error: calling lookup() on  NULL Dictionary refrence\n"); 
		exit(EXIT_FAILURE);
	}
	Node N = findKey(D,x);
	if( N !=NULL ){
		return N->value;
	}else{
		return NULL;
	} 
}


//insert()
void insert(Dictionary D, char* k, char* v){
   Node curr = NULL;
   Node new = newNode(k,v);

   if( D==NULL ){
                fprintf(stderr,
                "Dictionary Error: calling insert() on NULL Dictionary reference\n");
                exit(EXIT_FAILURE);
    } 
    

    Node temp = D->head;
    while(temp != NULL){
       if(k == temp->key){
               fprintf(stderr, 
               "Dictionary Error: cannot instert duplicate keys\n" );
                exit(EXIT_FAILURE);
       } 
      temp = temp->next;
   }
   if(isEmpty(D)){
     D->head=new;
   }else{
   for(curr=D->head; curr->next!=NULL; curr = curr->next ){
    }
       curr->next = new;
       }

   D->numItems++;
 return;
}



//delete 
void delete(Dictionary D, char* k){
	Node N;
	if(D==NULL){
		fprintf(stderr, "Dictionary Error: cannot delete on NULL Dictionary refrence.\n");
		exit(EXIT_FAILURE);
	}
	if(D->numItems==0){
		fprintf(stderr,"Dictionary Error: cannot delete empty dictionary.\n");
		exit(EXIT_FAILURE);
	}

        if(lookup(D, k) == NULL){
                fprintf(stderr,"Dictionary Error: cannot delete non-existant key.\n");
                exit(EXIT_FAILURE);
        }

	if(lookup(D, k) != NULL){
		if(D->numItems == 1){
			N = D->head;
			D->head = NULL;
			freeNode(&N);
			D->numItems--;
			return;
		}
		N = D->head;
		D->numItems--;
		if(strcmp(N->key,k) == 0){
			D->head = N->next;
			freeNode(&N);
		}else{
			N = D->head->next;
			Node P = D->head;
			while(N != NULL){
				if(strcmp(N->key,k) == 0){
					P->next = N->next;
					freeNode(&N);
					return;
				}else{ 
					P = N;
					N = N->next;
				}
			}
		}
	}
}

// printDictionary
// prints a text representation of S to the file pointed by out
// pre: none
void printDictionary(FILE* out, Dictionary D){
	Node N;
	if( D==NULL ){
		fprintf(stderr, 
				"Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	for(N=D->head; N!=NULL; N=N->next) fprintf(out, "%s %s\n", N->key, N->value);
	//fprintf(out, "\n");
}

//deleteAll
//deletes all pairs of string and value in dictionary
void makeEmpty(Dictionary D){
	if( D==NULL ){
		fprintf(stderr, 
				"Stack Error: calling popAll() on NULL Stack reference\n");
		exit(EXIT_FAILURE);
	}
	if( D->numItems==0 ){
		fprintf(stderr, "Stack Error: calling popAll() on empty Stack\n");
		exit(EXIT_FAILURE);
	}
	Node N = D->head;
	Node P;
	while(N != NULL){
		P = N;
		N = N->next;
		delete(D, P->key);
		//freeNode(&N);      
	}  
	D->numItems=0;
	//freeNode(&N);
	

}

// freeDictionary()
// destructor for the Stack type
void freeDictionary(Dictionary* pD){
	if( pD!=NULL && *pD!=NULL ){
		if( !isEmpty(*pD) ) makeEmpty(*pD);
		free(*pD);
		*pD = NULL;
	}
}

// size()
// returns the number of pairs in Dictionary
int size(Dictionary D){
	return D->numItems;
}




